pep-9999.rst

```diff
@@ -0,0 +1,520 @@
+released series, and preparing for the first beta release in the next
+release series.
+
+In even-numbered years, the focus for both events would just be on the upcoming
+maintenance and beta releases. These less intense years would hopefully provide
+an opportunity to tackle various process changes and infrastructure upgrades
+without impacting the release candidate preparation process.
+
+
+Release cycle alignment for prominent Linux distributions
+---------------------------------------------------------
+
+Some rolling release Linux distributions (e.g. Arch, Gentoo) may be in a
+position to consume the new rolling beta releases proposed in this PEP, but it
+is expected that most distributions would continue to use the established
+releases.
+
+The specific dates for those releases proposed in this PEP are chosen to align
+with the feature freeze schedules for the annual October releases of the Ubuntu
+and Fedora Linux distributions.
+
+For both Fedora and Ubuntu, it means that the release candidate phase aligns
+with the development period for a distro release, which is the ideal time for
+them to test a new version and provide feedback on potential regressions and
+compatibility concerns.
+
+For Ubuntu, this also means that their April LTS releases will have benefited
+from a full short-term release cycle using the new system Python version, while
+still having that CPython release be open to upstream bug fixes for most of the
+time until the next Ubuntu LTS release.
+
+The one Linux release cycle alignment that is likely to be consistently poor
+with the specific proposal in this PEP is with Debian, as that has been released
+in the first half of odd-numbered years since 2005 (roughly 12 months offset
+from Ubuntu LTS releases).
+
+With the annual release proposal in PEP 602, both Debian and Ubuntu LTS would
+consistently get a system Python version that is around 6 months old, but
+would also consistently select different Python versions from each other.
+
+With a two year cadence, and CPython releases in the latter half of the year,
+they're likely to select the same version as each other, but one of them will
+be choosing a CPython release that is more than 18 months behind the latest beta
+releases by the time the Linux distribution ships.
+
+If that situation does occur, and is deemed undesirable (but not sufficiently
+undesirable for *Debian* to choose to adjust their release timing), then that's
+where the additional complexity of the "incremental feature release" proposal
+in PEP 598 may prove worthwhile.
+
+(Moving CPython releases to the same half of the year as the Debian and Ubuntu
+LTS releases would potentially help mitigate the problem, but also creates
+new problems where a slip in the CPython release schedule could directly affect
+the release schedule for a Linux distribution)
+
+
+
+Implications for simple deployment environments
+-----------------------------------------------
+
+TODO ()
```

Comment:
May I suggest: "assuming a single person managing their own deployment"
* responsible for their own testing
* likely (or we recommend) consuming container images under CPython control
* able to choose their upgrade schedule

Reply:
Would you be able to take a shot at drafting these? I'm out of time for tonight, and am busy tomorrow night, so it'll be a couple of days before I get to it if we leave it up to me.

The corresponding sections that I have in PEP 598 can't be used directly, as what counts as "simple" and what counts as "complex" changes due to the fact that beta releases don't have a stable C ABI to target, but also don't have any expectations of universal forwards or backwards pickle compatibility.

Reply:
Your previous section for this part is basically okay. I tweaked it a little, but I think the "no implications" message is correct. (The other section will take more effort...)

> For the purposes of this PEP, a "simple" deployment environment is any use case where it is straightforward to ensure that all target environments are updated to a new Python release at the same time (or at least in advance of the rollout of new higher level application versions), and any pre-release testing that occurs need only target a single Python micro version.
>
> The simplest such case would be scripting for personal use, where the testing and target environments are the exact same environment.
>
> Similarly simple environments would be containerised web services, where the same Python container is used in the CI pipeline as is used on deployment, and any application that bundles its own Python runtime, rather than relying on a pre-existing Python deployment on the target system.
> 
> For these use cases, this PEP should not have any significant implications - only a single micro version needs to be tested, independently of whether that version is feature complete or not.
>
> The greatest point of additional complexity are "roll-back" compatibility with independently versioned features, such as pickle and SQLite, where use of new or provisional features in the beta stream may create files that are not readable by the stable release. Applications requiring the ability to roll-back to an early CPython release would, as today, be advised to avoid adopting prerelease versions.

Reply:
The big part my previous text didn't cover was the consequences of "the C ABI might change without notice". So I've kept the first 3 paragraphs as you suggested, along with the last paragraph about roll-back support, but reworked the middle part to talk about extension modules.